development

LiorRabin opened the pull request.

Commit:
basic ValidatorSet contract & tests

Commit:
renamed ValidatorSet to Consensus and improved basic tests

Commit:
more tests

Commit:
moved setSystemAddress to mock contract

Commit:
ValidatorState holds array of indexes

Commit:
add "stakeAmount.div(minStake)" times

Commit:
withdraw implementation and tests

Commit:
deployment, example configurations and readme

Commit:
flatten

leonprou commented on truffle.js:
```diff
@@ -0,0 +1,57 @@
+  },
+  mocha: {
+    reporter: 'eth-gas-reporter',
+    reporterOptions : {
```
extra space :)

leonprou commented on package.json:
```diff
@@ -0,0 +1,30 @@
+    "ethereumjs-wallet": "^0.6.3",
+    "ganache-cli": "^6.4.3",
+    "openzeppelin-solidity": "2.0.1",
+    "truffle": "4.1.14",
```
tuffle 5 got some nice features like choosing any solc version, or function overloading :)

leonprou commented on package.json:
```diff
@@ -0,0 +1,30 @@
+{
+  "name": "fuse-pos-network",
+  "version": "1.0.0",
```
better downgrade to 0.1.0 if we want semver to be useful

leonprou commented on migrations/2_deploy_contract.js:
```diff
@@ -0,0 +1,31 @@
+  }
+}
+
+// ./node_modules/.bin/truffle migrate --reset --network fuse_pos
```
please delete comment

leonprou commented on README.md:
```diff
@@ -1,2 +1,117 @@
+	* `echo "<THE_PASSWORD>" > pwd`
+* Edit `config.toml`
+	*  `[account]` section
+		*  `passowrd = ["pwd"]`
```
password*

leonprou commented on README.md:
```diff
@@ -1,2 +1,117 @@
+* Launch `node.0` to make sure everything is valid
+	* `parity --config config.toml`
+	* You should see a similar line every 5 seconds: `Imported #1 0x1204…4d39 (0 txs, 0.00 Mgas, 1 ms, 0.57 KiB)`
+	* Copy the public node address `enode://752963538fbf4fd29bef9845088763c93bfc9663bf2b4a5dd38408c3c55e0125fddea2d8c6af8a9182cf56d4e5114cf064ea89f225eb07abc026ac42a4404abb@40.0.0.67:30300` and save it to `enode`
```
2 quesitons
- I guess the public node address for me will be different, how to get it?
- what you mean by save it to `enode`? is it a file?

LiorRabin replied:
1. You see the address in the log of parity after you start it
2. `enode` is a file name, same like `pwd`
I will explain in the README

leonprou commented on README.md:
```diff
@@ -1,2 +1,117 @@
+* Edit `.env`
+	* `WALLET_PROVIDER_METHOD=keystore`
+	* `CREDENTIALS_KEYSTORE=~/Dev/io.parity.ethereum/alice/data/keys/FuseNetworkPOS/UTC--2019-04-30T11-43-32Z--a72d124a-5c05-c97c-e345-65c030649352`
+	* `CREDENTIALS_PASSWORD=/Users/liorrabin/Dev/io.parity.ethereum/alice/pwd`
```
change /Users/liorrabin/Dev/ to ~/Dev

leonprou commented on README.md:
```diff
@@ -1,2 +1,117 @@
+#### Deploy Consensus contract by **`alice`**
+* `cd ~/Dev/fuse-pos-network/`
+* `cp .env.example .env`
+* Edit `.env`
```
maybe it's more clean to just write all the changes as one block:
```
ENV1='..'
ENV2='..'
```

leonprou commented on README.md:
```diff
@@ -1,2 +1,117 @@
+* `cp .env.example .env`
+* Edit `.env`
+	* `WALLET_PROVIDER_METHOD=keystore`
+	* `CREDENTIALS_KEYSTORE=~/Dev/io.parity.ethereum/alice/data/keys/FuseNetworkPOS/UTC--2019-04-30T11-43-32Z--a72d124a-5c05-c97c-e345-65c030649352`
```
the user will have a different keystore file name

LiorRabin replied:
of course, it's just an example like all other names

leonprou commented on contracts/Consensus.sol:
```diff
@@ -0,0 +1,180 @@
+  address public SYSTEM_ADDRESS = 0xffffFFFfFFffffffffffffffFfFFFfffFFFfFFfE;
+
+  bool public finalized = false;
+  address public owner;
```
you can use openzeppelin [ownable](https://github.com/OpenZeppelin/openzeppelin-solidity/tree/master/contracts/ownership) contract

leonprou commented on contracts/Consensus.sol:
```diff
@@ -0,0 +1,180 @@
+
+  constructor(uint256 _minStake) public {
+    owner = msg.sender;
+    require (_minStake > 0);
```
better to move the require to `setMinStake`

leonprou commented on contracts/Consensus.sol:
```diff
@@ -0,0 +1,180 @@
+
+    stakeAmount[msg.sender] = stakeAmount[msg.sender].sub(_amount);
+
+    if(_amount >= minStake) {
```
if minStake = 1000, and I staked 1001 coins I''ll get 1 validator. Then I withdraw 999 coins, and will still have one validator cause 999 < 1000. 

Then I'll do this process again 1000 times and I'll get 1000 validators, staking 2 coins in each one :sweat_smile:

LiorRabin replied:
True :) will fix

leonprou commented on contracts/Consensus.sol:
```diff
@@ -0,0 +1,180 @@
+  uint256 public minStake;
+
+  address[] public currentValidators;
+  address[] public pendingValidators;
```
`pendingValidators` sound like all the validators there are in the pending state when actually they are the validators for the next block. maybe rename in to nextValidators  or fututeValidators

LiorRabin replied:
it's called `pending` on all examples/projects I've seen so far, so I kept the same convention

Commit:
minor fix

Commit:
use truffle 5

Commit:
PR comments

Commit:
use openzeppelin ownable

Commit:
remove truffle-wallet-provider

Commit:
updated truffle-hdwallet-provider after breaking change when migrating to truffle v5

Commit:
added ValidatorSet interface

Commit:
Consensus constructor adds an initial validator

Commit:
docker (#2)

Commit:
updated quickstart script

LiorRabin committed:
proxy implementation and test updates accordingly (#3)

Commit:
added bootnode role to docker, added bootnodes file, updated readme

Commit:
added role selection in quick-start script

Commit:
updated port to match azure security group

Commit:
updated zeronet bootnodes

Commit:
minor

Commit:
renamed repo

Commit:
Merge branch 'master' into dev

LiorRabin committed:
block reward (#4)

* basic block reward contract implementation and tests

* block reward contract deployment and network spec update

Commit:
added clean docker script

Commit:
storage refactor and voting implementation (#5)

Commit:
fix local ganache deployment

bejavu commented on contracts/BlockReward.sol:
```diff
@@ -0,0 +1,40 @@
+    address[] memory receivers = new address[](benefactors.length);
+    uint256[] memory rewards = new uint256[](receivers.length);
+
+    receivers[0] = benefactors[0];
```
assuming that `benefactors.length == 1`

LiorRabin replied:
Didn't understand the comment :)

bejavu commented on contracts/Consensus.sol:
```diff
@@ -0,0 +1,142 @@
+
+  function initialize(uint256 _minStake, address _initialValidator) public returns(bool){
+    require(!isInitialized());
+    setSystemAddress();
```
I would put the hard coded system address here as a parameter instead of inside the function.
Or maybe as a const at the beginning of the contract.

bejavu commented on contracts/BallotsStorage.sol:
```diff
@@ -0,0 +1,109 @@
+    return boolStorage[keccak256(abi.encodePacked("isInitialized"))];
+  }
+
+  function getBallotLimitPerValidator() public view returns(uint256) {
```
would love to see some code documentation, I don't think that I understand this indicator.

LiorRabin replied:
This function returns the number of "open" (active) ballots each validator (someone with voting rights) can have at the same time.

bejavu replied:
And what it supposed return value?

LiorRabin replied:
There's a default value of 100 max open ballots at a time (`getMaxLimitBallot` function).
If there are validators, this function will return 100/number_of_validators.
We can decide we don't need any limits and remove this functionality totally.

bejavu reviewed (changes requested):
Please add some code documentation

LiorRabin commented:
> Please add some code documentation

I will open a separate ticket for this, with a separate pull request

Commit:
pass system address as a parameter to the set storage function

Commit:
contracts documentation

LiorRabin commented:
@bejavu I've added contracts documentation

LiorRabin committed:
cycles and snapshots (#6)

* iniital implementation of cycles and snapshots

* make existing tests work after implementation changes and document new functions

* cycles and snapshots tests almost done

* added advance time helper

* cycle golden flow test

Commit:
refactor - all changes on network will be through contract implementation upgrade

[deleted all voting contracts]
[combined storage and logic functions into one contract]
[updated scripts and tests accordingly]

Commit:
changed cycle duration to be blocks instead of time

LiorRabin committed:
voting refactor (#7)

* replaced times with blocks (cycles)

* everyone can vote (save choice)

* count votes and finalize ballots on cycle end

* added interfaces for Consensus and Voting (needed to prevent circular contract dependencies)

* remove from active ballots array when ballot is finalized

* no message

LiorRabin committed:
simple gas optimizations (#8)

[change public function to external when possible]
[set storage keys as constants]
[delete variables after usage]
[no need to initialize simple types with default value]

LiorRabin committed:
delegators (#9)

* no message

* added stake function (payable)

* truffle v5.0.21

* delegation and withdrawal implemented

Commit:
deployment related updates

Commit:
fix number of snapshots taken each cycle

Commit:
fixed bug: validators were added to previous snapshot instead of overriding it

Commit:
added script to get ABIs from builds using jq

Commit:
added explorer_node role to parity wrapper

Commit:
consensus flow fixes

Commit:
revert part of previous commit

Commit:
some changes

LiorRabin committed:
refactor (#10)

* lint

* truffle v5.0.24

* added emitInitiateChange function

* work in progress

* work in progress

* get new validator set from snapshot logic implemented and tested for basic case

* better test

* updated golden flow test for new validator set from snapshot

* new validator set "shuffled"

* fixed case when only 1 validator

Commit:
get abi from build using jq installed by node-jq

LiorRabin committed:
dynamic block reward (yearly inflation) (#11)

* BlockReward initialize params changed to initial_supply and yearly-inflation-percentage (reward amount to be calculated dynamically)

* implemented dynamic block reward amount calculation

* test updating block reward amount after "blocks_per_year" and correct yearly inflation in total supply

* fixed consensus test

Commit:
renamed voting test file

LiorRabin committed:
off-chain app for validators (#12)

* created node app which will run as another docker (for validators only)

* work in progress

* work in progress

* done

* send transaction with zero gas price

Commit:
quickstart script fix for validator app container

Commit:
adjust truffle migrate

Commit:
stringify error

Commit:
gas limit

Commit:
removed `participant` role

LiorRabin committed:
refactor changing validator set on cycle end (#13)

* some settings

* implementation (including some contract size optimizations)

* some fixes

* implementation

Commit:
consensus - on cycle end - random snapshot of pending validators is selected to be the new set

LiorRabin committed:
documentation (#14)

* update project readme

* contracts description

* app readme

LiorRabin committed:
review comments (#15)

* voting - removed total voters

* consensus - remove from pending validators when less than min stake

* consensus - stake is actually delegate to myself (solves issue of validator withdrawing staked amount by a delegator)

Commit:
quickstart fix

Commit:
updated docker hub organization & repos

Commit:
added hardware specs

Commit:
test

bejavu reviewed (approved):
Great job!
